### PR TITLE
Update 4_compile.py

### PR DESCRIPTION
**Pull request by msaroufim**
Haven't ran this yet but included 2 important caveats that will have a big performance implication - will run this next and see what else we can debug

### CODE REVIEW

File: 4_compile.py [hunk 1]
```diff
@@ -112,6 +112,10 @@ def configure_optimizers(self):
 if __name__ == "__main__":
     print(watermark(packages="torch,lightning,transformers", python=True), flush=True)
     print("Torch CUDA available?", torch.cuda.is_available(), flush=True)
+    
+    # Enable Tensor Cores
+    if torch.cuda.get_device_capability() >= (8, 0):
```
Comment (msaroufim) [inline 4_compile.py:117]: This is probably the single most important flag to configure, we can't enable it by default since it changes the numerics of a program but at this point I am heavily considering enabling it by default

File: 4_compile.py [hunk 2]
```diff
@@ -169,7 +173,8 @@ def configure_optimizers(self):
     model = AutoModelForSequenceClassification.from_pretrained(
         "distilbert-base-uncased", num_labels=2
     )
-
+    
+    # First model inference will trigger compilation
     model = torch.compile(model)
```
Comment (msaroufim) [inline 4_compile.py:178]: Passing in a sample batch first might help or ensuring the training run is long enough so compilation time is amortized

**Comment by rasbt** [comment]
Awesome, thanks for jumping in here. Would love to get some insights wrt to how to improve that. I should mentioned, I used CUDA 11.8. 

Let me try the sample batch idea!

File: 4_compile.py (hunk 1 above)
Comment (rasbt) [reply 4_compile.py:117]: what would you insert into this context to enable the tensor cores explicitely?

Comment (msaroufim) [reply 4_compile.py:117]: Take a look at this https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html

Basically high means do all computations with tf32 or bf16 which is what tensor cores do

A more readable but less good version of the above is torch.backends.cuda.matmul.allow_tf32 = True

Comment (rasbt) [reply 4_compile.py:117]: Thanks, but for ampere GPUs, it (=usage of tensor cores) should happen automatically I thought?

File: 4_compile.py (hunk 2 above)
Comment (rasbt) [reply 4_compile.py:178]: Weird question, when I use DeepSpeed, I would have to do that on each device separately?

File: 4_compile.py (hunk 1 above)
Comment (msaroufim) [reply 4_compile.py:117]: That's unfortunately not the case by default, but you might be using them if you use cast your tensors to bf16 or tf32 or if you use AMP 

I find it helpful to run NVIDIA Nsight and look at their tensor core utilization there to make sure

Comment (rasbt) [reply 4_compile.py:117]: Oh this code is using AMP. I meant in an AMP context above of course. Sorry if I wasn't clear.

**Comment by msaroufim** [comment]
Ah your batch size is also quite small so might be best to try out `torch.compile(m, mode="reduce-overhead")` which will automatically enable cuda graphs for you

Recently added some docs to make most of this clearer https://pytorch.org/docs/master/compile/index.html

**Comment by rasbt** [comment]
> Ah your batch size is also quite small so might be best to try out `torch.compile(m, mode="reduce-overhead")` which will automatically enable cuda graphs for you
> 
> Recently added some docs to make most of this clearer https://pytorch.org/docs/master/compile/index.html

Thanks, I tried that originally and it didn't really help :(.

**Comment by msaroufim** [comment]
Yeah should be in combination with tensor cores

File: 4_compile.py (hunk 1 above)
Comment (msaroufim) [reply 4_compile.py:117]: Ah! Well let's see if you get any changes, otherwise might need to start profiling to see what's going on

File: 4_compile.py (hunk 2 above)
Comment (rasbt) [reply 4_compile.py:178]: Nice, I was actually able to shave of a minute by doing that. 

Now, the only problem is that the model doesn't work for inference when I load it from a checkpoint. Any ideas for how to fix that?

Comment (rasbt) [reply 4_compile.py:178]: So what I did was 

```python
    model.to(torch.device("cuda:0"))
    model = torch.compile(model)

    for batch_idx, batch in enumerate(train_loader):
        model.train()
        for s in ["input_ids", "attention_mask", "label"]:
            batch[s] = batch[s].to(torch.device("cuda:0"))
        break

    outputs = model(
        batch["input_ids"],
        attention_mask=batch["attention_mask"],
        labels=batch["label"],
    )

    lightning_model = LightningModel(model)
    # start timing and training below
```

This was for a single-GPU context. But how would I do that in a multi-GPU context where each GPU keeps its own model. Would I have to compile it on each device or is it okay to compile it on the first device? Any recommendations for DeepSpeed contexts?

Comment (rasbt) [reply 4_compile.py:178]: After training, is there any command required to make it compatible with inference mode perhaps so that I can evaluate the model on the test set?

Comment (msaroufim) [reply 4_compile.py:178]: If by inference mode you mean https://pytorch.org/docs/stable/generated/torch.inference_mode.html then our suggestion would be not to use it anymore. For evaluation standard no_grad should be sufficient

As far as compiling per device, my understanding is yes you would need to do that @wconstab could give you more recommendations on the distributed front. Inductor does maintain a code cache and it doesn't seem all that crazy to copy it over but I don't think we do that. Keep in mind that distributed support is still early and you can follow updates here https://github.com/pytorch/pytorch/issues/93173

Regarding issues on loading from a checkpoint the main annoying thing is you have to keep in mind that an OptimizedModule is not an nn.Module (we're discussing making some changes here) so you might to check how you're loading or saving a model, see this https://pytorch.org/get-started/pytorch-2.0/#serialization

Comment (carmocca) [reply 4_compile.py:178]: @msaroufim Can you elaborate on this comment?

> our suggestion would be not to use it anymore

Do you mean in general or in the context of `torch.compile`? Why would it not be suggested anymore?

Comment (msaroufim) [reply 4_compile.py:178]: Only in the context of torch.compile
